packages/repository/src/decorators/relation.repository.decorator.ts

```diff
@@ -0,0 +1,59 @@
+import {Entity} from '../model';
+import {EntityCrudRepository} from '../repositories/repository';
+import {Class} from '../common-types';
+import {
+  HasManyDefinition,
+  hasManyRepositoryFactory,
+  RelationDefinitionBase,
+  getConstrainedRepositoryFunction,
+} from '../repositories/relation.factory';
+import {RelationType, RELATIONS_KEY} from './relation.decorator';
+import {
+  inject,
+  BindingKey,
+  Context,
+  Injection,
+  MetadataInspector,
+} from '@loopback/context';
+import {DefaultCrudRepository} from '../repositories/legacy-juggler-bridge';
+import {DataSource} from 'loopback-datasource-juggler';
+
+/**
+ * Decorator for hasMany
+ * @param targetRepo
+ * @returns {(target:any, key:string)}
+ */
+export function hasManyRepository<T extends Entity>(
+  targetRepo: Class<EntityCrudRepository<T, typeof Entity.prototype.id>>,
+) {
+  // tslint:disable-next-line:no-any
+  return function(target: any, key: string, index: number) {
+    inject(
+      BindingKey.create<typeof targetRepo>(`repositories.${targetRepo.name}`),
+      {sourceRepo: target},
+      resolver,
+    )(target, key, index);
+  };
+
+  async function resolver(ctx: Context, injection: Injection) {
+    const tRepo = await ctx.get<
+      DefaultCrudRepository<Entity, typeof Entity.prototype.id>
+    >(injection.bindingKey);
+
+    /**
+     * discussion point: use a fake DataSource instance so that we can get back
+     * the source model possible alternatives we considered:
+     *  - use context to resolve an instance of the source repository and thus a
+     *    source model (results in circular dependency between related
+     *    repositories)
+     *  - change UX so that users pass in the source model manually
+     *    (inconvenient)
+     *  - enforce binding of models to application context, so that it can be
+     *    retrieved
+     */
+
+    const fakeDs = new DataSource();
```

Comment:
?

Reply:
The reasoning for this is listed above the line here.
What do you think about the dilemma we may have at hand here?

Reply:
I don't understand what problem you are trying to solve, could you please explain in more detail?

Why does `getConstrainedRepositoryFunction` need a full `sourceModel` (a PersistedModel)? IIUC, you are trying to find the `@hasMany` decorator on the source model. Don't we already have that information available here inside `hasManyRepository` decorator?

I think you need to rethink the design of new decorators and hasManyRepositoryFactory function.

To me, the need to create a fake datasource is strong signal that there is something wrong with the currently proposed design.

Reply:
> IIUC, you are trying to find the @hasMany decorator on the source model. 

Correct.

>Don't we already have that information available here inside hasManyRepository decorator?

Yes, and the way we saw it, we had to instantiate a new instance of the `target` repository class that the decorator has access to in order to get the source model. Is there another way you're thinking of?

>I think you need to rethink the design of new decorators and hasManyRepositoryFactory function.

>To me, the need to create a fake datasource is strong signal that there is something wrong with the currently proposed design.

Yeah I agree, and that's why we laid out alternatives too. I think your proposal to inject the datasource for the related model and construct the target repository that way would work well 👍.